src/Transports.AspNetCore/WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,253 @@
+    public CancellationToken RequestAborted { get; }
+
+    /// <summary>
+    /// Returns the default disconnection timeout value.
+    /// See <see cref="GraphQLWebSocketOptions.DisconnectionTimeout"/>.
+    /// </summary>
+    protected virtual TimeSpan DefaultDisconnectionTimeout { get; } = TimeSpan.FromSeconds(10);
+
+    /// <inheritdoc/>
+    public DateTime LastMessageSentAt { get; private set; } = DateTime.UtcNow;
+
+    /// <inheritdoc/>
+    public HttpContext HttpContext { get; }
+
+    /// <summary>
+    /// Initializes an instance with the specified parameters.
+    /// </summary>
+    public WebSocketConnection(HttpContext httpContext, WebSocket webSocket, IGraphQLSerializer serializer, GraphQLHttpMiddlewareOptions options, CancellationToken cancellationToken)
+    {
+        HttpContext = httpContext ?? throw new ArgumentNullException(nameof(httpContext));
+        if (options == null)
+            throw new ArgumentNullException(nameof(options));
+        if (options.WebSockets.DisconnectionTimeout.HasValue)
+        {
+            if ((options.WebSockets.DisconnectionTimeout.Value != Timeout.InfiniteTimeSpan && options.WebSockets.DisconnectionTimeout.Value.TotalMilliseconds < 0) || options.WebSockets.DisconnectionTimeout.Value.TotalMilliseconds > int.MaxValue)
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly
+                throw new ArgumentOutOfRangeException(nameof(options) + "." + nameof(GraphQLWebSocketOptions.DisconnectionTimeout));
+#pragma warning restore CA2208 // Instantiate argument exceptions correctly
+        }
+        _closeTimeoutMs = (int)(options.WebSockets.DisconnectionTimeout ?? DefaultDisconnectionTimeout).TotalMilliseconds;
+        _webSocket = webSocket ?? throw new ArgumentNullException(nameof(webSocket));
+        _stream = new(webSocket);
+        _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
+        _pump = new(HandleMessageAsync);
+        RequestAborted = cancellationToken;
+    }
+
+    /// <summary>
+    /// Listens to incoming messages on the WebSocket specified in the constructor,
+    /// dispatching the messages to the specified <paramref name="operationMessageProcessor"/>.
+    /// Returns or throws <see cref="OperationCanceledException"/> when the WebSocket connection is closed.
+    /// </summary>
+    public virtual async Task ExecuteAsync(IOperationMessageProcessor operationMessageProcessor)
+    {
+        if (operationMessageProcessor == null)
+            throw new ArgumentNullException(nameof(operationMessageProcessor));
+        if (Interlocked.Exchange(ref _executed, 1) == 1)
+            throw new InvalidOperationException($"{nameof(ExecuteAsync)} may only be called once per instance.");
+        try
+        {
+            await operationMessageProcessor.InitializeConnectionAsync();
+            // set up a buffer in case a message is longer than one block
+            var receiveStream = new MemoryStream();
```

Comment:
Do you have any questions here?  Yes it is complicated to read.  Yes it would certainly be more optimized with the MS package you previously referenced.  But everything should work.

Note that I do not know how likely it is for a message to come through as a single packet (and so retain in the 16kb buffer).  It is possible that messages often come through as multiple packets, especially if TCP breaks apart packets over 1500 bytes.  This could be further optimized by reusing the 16kb buffer until it was full before switching to the stream.  

It's also possible that the message comes through as a packet and then the end-of-message packet comes through separately (just as we send messages).  This would currently hit the stream instead of staying within the 16kb buffer (without further optimization).

Reply:
No suggestions here. I'm fine with current implementation.

Reply:
#813